src/coreclr/gc/gc.cpp

```diff
@@ -8368,29 +8390,247 @@ void destroy_card_table (uint32_t* c_table)
     dprintf (2, ("Table Virtual Free : %Ix", (size_t)&card_table_refcount(c_table)));
 }
 
+void gc_heap::get_card_table_element_sizes (uint8_t* start, uint8_t* end, size_t sizes[total_bookkeeping_elements])
+{
+    memset (sizes, 0, sizeof(size_t) * total_bookkeeping_elements);
+    sizes[card_table_element] = size_card_of (start, end);
+    sizes[brick_table_element] = size_brick_of (start, end);
+#ifdef CARD_BUNDLE
+    if (can_use_write_watch_for_card_table())
+    {
+        sizes[card_bundle_table_element] = size_card_bundle_of (start, end);
+    }
+#endif //CARD_BUNDLE
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    if (gc_can_use_concurrent)
+    {
+        sizes[software_write_watch_table_element] = SoftwareWriteWatch::GetTableByteSize(start, end);
+    }
+#endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    sizes[seg_mapping_table_element] = size_seg_mapping_table_of (start, end);
+    sizes[mark_array_element] = 0;
+#ifdef BACKGROUND_GC
+    if (gc_can_use_concurrent)
+    {
+        sizes[mark_array_element] = size_mark_array_of (start, end);
+    } 
+#endif //BACKGROUND_GC
+}
+
+void gc_heap::get_card_table_element_layout (uint8_t* start, uint8_t* end, size_t layout[total_bookkeeping_elements + 1])
+{
+    size_t sizes[total_bookkeeping_elements];
+    get_card_table_element_sizes(start, end, sizes);
+
+    layout[card_table_element] = sizeof(card_table_info);
+    layout[brick_table_element] = layout[card_table_element] + sizes[card_table_element];
+#ifdef CARD_BUNDLE
+    layout[card_bundle_table_element] = layout[brick_table_element] + sizes[brick_table_element];
+#endif
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    layout[software_write_watch_table_element] = layout[software_write_watch_table_element - 1] + sizes[software_write_watch_table_element - 1];
+    if (gc_can_use_concurrent)
```

Comment:
is there any reason why you check gc_can_use_concurrent after you've filled in the layout for software ww? it's not needed if gc_can_use_concurrent is false.

Reply:
`layout[software_write_watch_table_element] == layout[seg_mapping_table_element - 1]` when `FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP` is defined, so the value is read on line 8436 and therefore must be filled on line 8430 even when background GC is disabled.

There is no need to waste space for alignment in case the background GC is disabled.